### PR TITLE
[Components]: Add `ToggleGroupControlOptionIcon` component

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Part of: https://github.com/WordPress/gutenberg/issues/36230
Needed for: https://github.com/WordPress/gutenberg/issues/36735 

This PR adds `icon` support to `ToggleGroupControlOption` component (child of ToggleGroupControl). 

## Why?
This will be needed for implementing other controls like `textTransform` or `textDecoration` to use `ToggleGroupControl`.

## How?
This implementation makes the use of icon and label mutually exclusive. What this means is that if we provide an `icon` it will be used instead of the `label` and a tooltip will be shown automatically.

## Testing Instructions
1. Check the storybook examples of the component.

## Notes
I've noticed that in storybook the options are not set properly when we switch to different stories. It's not related to this PR and it should be handled separately.

https://user-images.githubusercontent.com/16275880/160121986-9ae847b8-ae44-4aca-910f-2d9dbad92bf9.mov

## Screenshots or screencast 

https://user-images.githubusercontent.com/16275880/160121691-ea7652ff-96f2-44b1-8c7a-e6e1f655a311.mov

### CODE REVIEW

File: packages/components/src/toggle-group-control/types.ts
Comment (ntsekouras) [inline packages/components/src/toggle-group-control/types.ts]: I'm not sure about the best type here... --cc @ciampo

Comment (ramonjd) [reply packages/components/src/toggle-group-control/types.ts]: `JSX.Element` is the type I get back from the compiler for components in `icons/src/library/*`

Screen Shot 2022-03-28 at 11 54 23 am

**Review by ramonjd — APPROVED** [review]
This LGTM, but I'd defer to @mirka or @ciampo for their opinions.

I thought an update to the storybook npm packages might fix the args not updating on the canvas, but it didn't for me. Possibly related issue: https://github.com/storybookjs/storybook/issues/16174

File: packages/components/src/toggle-group-control/toggle-group-control-option/styles.ts
Comment (mirka) [inline packages/components/src/toggle-group-control/toggle-group-control-option/styles.ts]: For overall consistency, I think we should do `fill: currentColor` inside this block instead:

https://github.com/WordPress/gutenberg/blob/d89572e7716cdcc07ffaff129f8259be5ae2d477/packages/components/src/toggle-group-control/toggle-group-control-option/styles.ts#L23

This will match the text color, and also get any other state-based color overrides.

File: packages/components/CHANGELOG.md
Comment (mirka) [inline packages/components/CHANGELOG.md]: We should also add the prop in the [readme](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/toggle-group-control/toggle-group-control-option/README.md).

**Review by mirka — COMMENTED** [review]
Super happy to see this getting closer to unification! Pinging @jasmussen for a design review, since it's not entirely clear to me from the Figma mockups what the intended proportions are. (Mostly talking about the horizontal paddings. Should it be narrower so each icon option is a square? It kind of needs to be narrower for it to fit in the 50% sidebar width like in the mockups.)

> in storybook the options are not set properly when we switch to different stories

My hunch is that this is due to using both Knobs and Controls in the same stories. I don't think they're designed to coexist. And even on their own, Knobs are pretty buggy when clicking between stories! We'll need to refactor them all to Controls soon.

File: packages/components/src/toggle-group-control/toggle-group-control-option/styles.ts
Comment (ntsekouras) [reply packages/components/src/toggle-group-control/toggle-group-control-option/styles.ts]: Good catch! Thanks!

**Comment by ntsekouras** [comment]
Thanks for the reviews!

>for a design review, since it's not entirely clear to me from the Figma mockups what the intended proportions are. (Mostly talking about the horizontal paddings. Should it be narrower so each icon option is a square? It kind of needs to be narrower for it to fit in the 50% sidebar width like in the mockups.)

@mirka I saw that padding because I was testing this with changing the current `textTransform` control and I think we can handle this in a follow up. This has to do with the current padding applied(`0 12px`) and affects the component in general.

I think when we'll make the transition to other typography controls we might end up with some small differences from the original mockup, and I don't think that's a bad thing.  It will be easier to test/iterate there.

**Comment by ciampo** [comment]
> We'll need to refactor them all to Controls soon.

Yup, we should definitely convert our Storybook examples to Controls!

> This implementation makes the use of icon and label mutually exclusive. What this means is that if we provide an icon it will be used instead of the label and a tooltip will be shown automatically.

I'm not sure if this is the best strategy for this component's APIs:

- `label` is required, but if using an icon it will be used for the tooltip instead
- if `aria-label` is passed, `label` is not used for the tooltip
- and so there is the change that the `label` prop is passed (as it's required), but it's never used in case both `icon` and `aria-label` are passed

We could also evaluate a different approach, which aims at simplifying the API by removing any hidden conventions:

- `label` is not required
- if both `label` and `icon` are passed, the component displays both (and therefore it would be up to the consumer of the component to pass only what is necessary)
- the `showTooltip` prop would need to be explicitly set (and therefore it would be up to the consumer of the component to set the prop and pass an appropriate `label` or, even better in case of icons, an `aria-label`)

What do folks think?

**Comment by jasmussen** [comment]
Thanks for the ping, nice one.

> Needed for: https://github.com/WordPress/gutenberg/issues/36735

I would caution this a bit, for two reasons. 1: we still need an interface control for a row of toggles, similar to bold/italic/link in the Paragraph toolbar. Here's a couple of examples:

Inspector

width


These are more like buttons or individual toggles, than they are items in a group. Furthermore, [you can combine both strike-through and underline](https://codepen.io/joen/pen/popwNwJ?editors=1100), so they aren't mutually exclusive, even if in the interface they are at the moment.

That's not to say we shouldn't add an icon to the option group, just that it likely can't replace the basic togglegroup.

> Super happy to see this getting closer to unification! Pinging @jasmussen for a design review, since it's not entirely clear to me from the Figma mockups what the intended proportions are. (Mostly talking about the horizontal paddings. Should it be narrower so each icon option is a square? It kind of needs to be narrower for it to fit in the 50% sidebar width like in the mockups.)

Yes indeed, it's changed a few times back and forth. Like with inputs, I suspect this is one of the cases where we'd want to support both 36 and 40. I see it at 32px in Global Styles, not sure we need that, but it seems fine to start with 36 and go from there.

#38990 and #38577 both use a basic segmented control and the metrics we could start with are these:

segmented

Screenshot 2022-03-29 at 08 59 18

* 36px by default
* 1px border, #949494
* 2px outer border radius (i.e. `border-radius: 2px;` if an inset shadow, `1px` if an outset border)
* 2px inner padding between the gray and the dark toggle
* 2px border radius on the pill inside

Let me know if that was helpful!

**Comment by ntsekouras** [comment]
> We could also evaluate a different approach, which aims at simplifying the API by removing any hidden conventions:
>- `label` is not required
>- if both `label` and `icon` are passed, the component displays both (and therefore it would be up to the consumer of the component to pass only what is necessary)
>- the `showTooltip` prop would need to be explicitly set (and therefore it would be up to the consumer of the component to set the prop and pass an appropriate `label` or, even better in case of icons, an `aria-label`)

@ciampo we could try that. Personally I think showing both icon and label will make this a bit crowded, but 🤷 .

Also if we make label optional, is it okay to allow none of them passed?

>both use a basic segmented control and the metrics we could start with are these:
>36px by default
1px border, #949494
2px outer border radius (i.e. border-radius: 2px; if an inset shadow, 1px if an outset border)
2px inner padding between the gray and the dark toggle
2px border radius on the pill inside

@jasmussen these styles are for the component in general, correct? If yes I can open a separate PR for that, as it's not related to this one.

**Comment by jasmussen** [comment]
> these styles are for the component in general, correct? If yes I can open a separate PR for that, as it's not related to this one.

Yes, the general component. My mistake!

**Comment by jasmussen** [comment]
> if both label and icon are passed, the component displays both (and therefore it would be up to the consumer of the component to pass only what is necessary)

I'd do icon or label, not both, exactly for the space problems mentioned. If we find a particular need for this combination in the future, we can always revisit, but I wouldn't start with this.

**Comment by ciampo** [comment]
> Personally I think showing both icon and label will make this a bit crowded, but 🤷

I agree that it will look crowded — in this scenario, it would be up to the consumer of `ToggleGroupControl` to pass only what is necessary (icon or text label).

> Also if we make label optional, is it okay to allow none of them passed?

It's not a great option, but my reasoning is that a situation where the component doesn't show anything (because neither `label` not `icon` were passed) is quite easy to spot and to understand what is the fix for.

If the content of the `label` prop is shown in different ways depending on other props (text label, toolptip text, or even ignored if `aria-label` and `icon` props are passed), it makes difficult for a consumer of `ToggleGroupControl` to understand what is happening and debug any issues.

**I am basically not a fan of component APIs that rely on conventions** like the ones proposed above, e.g "if `propA` is specified, than `propB` is ignored and `propC` is forced to `true`". It makes it hard to maintain and evolve a component, and difficult for a consumer to understand the APIs

Other alternative approaches could be:
- to stop relying on `label` and replace it with a generic `children` mandatory prop — at that point, we could pass any contents to `ToggleGroupControlOption`. We could still write styles for text and icons in `ToggleGroupControl` to make sure that the design specs are applied
- We could create a different `ToggleGroupControlIconOption` component with its own props

**Comment by ciampo** [comment]
I missed this comment when writing my latest reply (thank you @ntsekouras for pointing it out!)

> It's not a great option, but my reasoning is that a situation where the component doesn't show anything (because neither `label` not `icon` were passed) is quite easy to spot and to understand what is the fix for.

I agree that the approach that I proposed earlier (_if both `label` and `icon` are passed, the component displays both_) is not ideal, and I'm happy to drop it.

I left some thoughts about other alternative approaches in [my last comment](https://github.com/WordPress/gutenberg/pull/39760#issuecomment-1081625738), let me know your thoughts!

**Comment by mirka** [comment]
@jasmussen Talking specifically about the horizontal paddings, should we decrease them to make each icon button a tight square, or keep the current `ToggleGroupControlOption` paddings for a rectangular shape?

| Square | Rectangle |
|--------|-------|
|Mockup with square-shaped icon button|Implementation with rectangular icon button|

(No strong opinion either way. It might even make sense to keep them visually distinct like this, so it's easier to grasp the difference between a group that's mutually exclusive and one that's not.)

**Comment by mirka** [comment]
> these styles are for the component in general, correct? If yes I can open a separate PR for that, as it's not related to this one.

@ntsekouras Just FYI, the Components Squad will be coordinating height changes (#39397, though `ToggleGroupControl` is already correct) and border colors (#39425), so no need for you to independently address these two points for this specific component 👍

**Comment by mirka** [comment]
cc: @ciampo 

> * to stop relying on `label` and replace it with a generic `children` mandatory prop — at that point, we could pass any contents to `ToggleGroupControlOption`. We could still write styles for text and icons in `ToggleGroupControl` to make sure that the design specs are applied
> * We could create a different `ToggleGroupControlIconOption` component with its own props

My vote goes to `ToggleGroupControlIconOption`, which I think is the most straightforward. Which, interestingly, would best be implemented in combination with the first approach ("generic `children` mandatory prop").

For example, something like this could be backwards compatible and simple to maintain:

1. Internally, genericize `ToggleGroupControlOption` so it takes a `children` mandatory prop. Name it something different like `ToggleGroupControlOptionBase`. (We don't necessarily have to export this component publicly yet.)
2. Using `ToggleGroupControlOptionBase`, make a thin wrapper component called `ToggleGroupControlOption` and keep the existing APIs (i.e. forward the `label` prop to `<ToggleGroupControlOptionBase>{ label }</ToggleGroupControlOptionBase>`).
3. Using `ToggleGroupControlOptionBase`, make a thin wrapper component called `ToggleGroupControlIconOption` for icon buttons.

**Comment by jasmussen** [comment]
> It's not a great option, but my reasoning is that a situation where the component doesn't show anything (because neither label not icon were passed) is quite easy to spot and to understand what is the fix for.

If you provide both a label and an icon, can't we use the label just for aria-label?

**Comment by jasmussen** [comment]
> Talking specifically about the horizontal paddings, should we decrease them to make each icon button a tight square, or keep the current ToggleGroupControlOption paddings for a rectangular shape?

Ah, that's a good question. Visually, I'd actually think these work best the same as they are today:

same

That is — the black buttons are 32x32 inside a borderless 36px tall container.

Should we add a borderless version of the segmented control to account for this?

**Comment by ciampo** [comment]
> If you provide both a label and an icon, can't we use the label just for aria-label?

We could, but I see two drawbacks with this approach:

- the contents of the `label` prop would be used in two different ways depending on the value of other props (in normal conditions, the `label` contents would be displayed to the user — but if the `icon` prop is specified, the `label` contents would either be used for accessibility-only purposes, or moved to a tooltip)
- if the standard HTML `aria-label` prop is defined and the `icon` prop is defined, the contents of the `label` prop would be ignored

I'd personally prefer the approach highlighted by @mirka above, which is more straightforward in terms of API design and easier to maintain in the future:

> My vote goes to `ToggleGroupControlIconOption`, which I think is the most straightforward. Which, interestingly, would best be implemented in combination with the first approach ("generic `children` mandatory prop").
> 
> For example, something like this could be backwards compatible and simple to maintain:
> 
> 1. Internally, genericize `ToggleGroupControlOption` so it takes a `children` mandatory prop. Name it something different like `ToggleGroupControlOptionBase`. (We don't necessarily have to export this component publicly yet.)
> 2. Using `ToggleGroupControlOptionBase`, make a thin wrapper component called `ToggleGroupControlOption` and keep the existing APIs (i.e. forward the `label` prop to `<ToggleGroupControlOptionBase>{ label }</ToggleGroupControlOptionBase>`).
> 3. Using `ToggleGroupControlOptionBase`, make a thin wrapper component called `ToggleGroupControlIconOption` for icon buttons.

File: packages/components/src/toggle-group-control/stories/index.js
Comment (ciampo) [inline packages/components/src/toggle-group-control/stories/index.js]: Since `showTooltip` is `true`, let's add an `aria-label` prop (otherwise the tooltip won't be shown)

 ```suggestion
			<ToggleGroupControlOptionIcon
				value="uppercase"
				icon={ formatUppercase }
				showTooltip={ true }
				aria-label="Uppercase"
			/>
			<ToggleGroupControlOptionIcon
				value="lowercase"
				icon={ formatLowercase }
				showTooltip={ true }
				aria-label="Lowercase"
			/>
		</ToggleGroupControl>
```

File: packages/components/src/toggle-group-control/test/index.js
Comment (ciampo) [inline packages/components/src/toggle-group-control/test/index.js]: Same as before — if we pass `showTooltip={true}`, we should also add an `aria-label` so that the tooltip gets rendered

File: packages/components/src/toggle-group-control/toggle-group-control-option-base/README.md
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option-base/README.md]: Tiny nit

```suggestion
### `children`: `ReactNode`

The children elements.
```

File: packages/components/src/toggle-group-control/toggle-group-control-option-base/README.md
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option-base/README.md]:
```suggestion
`ToggleGroupControlOptionBase` is a form component and is meant to be used as an internal, generic component for any children of [`ToggleGroupControl`](<(/packages/components/src/toggle-group-control/toggle-group-control/README.md)>).
```

File: packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx]: I'd apply the same copy changes as the ones suggested in the `ToggleGroupControlOptionBase`'s README

File: packages/components/CHANGELOG.md
Comment (ciampo) [inline packages/components/CHANGELOG.md]: We should update the changelog mentioning that we're introducing a new `ToggleGroupControlOptionIcon` component

File: packages/components/src/toggle-group-control/toggle-group-control-option-base/README.md
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option-base/README.md]:
```suggestion
Whether to show a tooltip when hovering over the option. The tooltip will only show if a label for it is provided using the `aria-label` prop.
```

File: packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option/component.tsx]: We should mark `ToggleGroupControlOptionBase`, `ToggleGroupControlOption` and `ToggleGroupControlOptionIcon` as non-polymorphic, since they are internally always going to render `as="button"`

```suggestion
	props: WordPressComponentProps< ToggleGroupControlOptionProps, 'button', false >
```

File: packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx]: We should mark `ToggleGroupControlOptionBase`, `ToggleGroupControlOption` and `ToggleGroupControlOptionIcon` as non-polymorphic, since they are internally always going to render `as="button"`

```suggestion
	props: WordPressComponentProps<
		ToggleGroupControlOptionBaseProps,
		'button',
		false
	>,
```

File: packages/components/src/toggle-group-control/toggle-group-control-option-icon/component.tsx
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option-icon/component.tsx]: We should mark `ToggleGroupControlOptionBase`, `ToggleGroupControlOption` and `ToggleGroupControlOptionIcon` as non-polymorphic, since they are internally always going to render `as="button"`

```suggestion
	props: WordPressComponentProps<
		ToggleGroupControlOptionIconProps,
		'button',
		false
	>
```

File: packages/components/src/toggle-group-control/toggle-group-control-option-icon/README.md
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option-icon/README.md]:
```suggestion
Icon displayed as the content of the the `ToggleGroupControl` option.
```

File: packages/components/src/toggle-group-control/toggle-group-control-option-icon/README.md
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option-icon/README.md]:
```suggestion
Whether to show a tooltip when hovering over the option. The tooltip will only show if a label for it is provided using the `aria-label` prop.
```

File: packages/components/src/toggle-group-control/toggle-group-control-option-icon/README.md
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option-icon/README.md]: Let's also add an `aria-label` prop in this example so that the tooltip gets shown

**Review by ciampo — COMMENTED** [review]
Thank you so much @ntsekouras for applying the feedback!

I've left a few comments, but most of them a very minor requests. We're close to being able to merge this 🎉

**Comment by ntsekouras** [comment]
Thanks for the review and the help here @ciampo! Feedback addressed 😄

File: packages/components/src/toggle-group-control/toggle-group-control-option-icon/component.tsx
Comment (ntsekouras) [reply packages/components/src/toggle-group-control/toggle-group-control-option-icon/component.tsx]: Why we didn't have this before?

Comment (ciampo) [reply packages/components/src/toggle-group-control/toggle-group-control-option-icon/component.tsx]: Not sure. From a quick look, it seems that the component's prop were already like this since the component was called `SegmentedControl`. Probably just something that got overlooked?

File: packages/components/src/toggle-group-control/toggle-group-control-option-icon/README.md
Comment (ciampo) [inline packages/components/src/toggle-group-control/toggle-group-control-option-icon/README.md]:
```suggestion
Icon displayed as the content of the option. Usually one of the icons from the `@wordpress/icons` package, or a custom React `<svg>` icon.
```

File: packages/components/src/toggle-group-control/types.ts
Comment (ciampo) [inline packages/components/src/toggle-group-control/types.ts]:
```suggestion
	 * Icon displayed as the content of the option. Usually one of the icons from
	 * the `@wordpress/icons` package, or a custom React `<svg>` icon.
```

**Review by ciampo — APPROVED** [review]
Thank you for your patience when working on this PR, @ntsekouras !

Changes LGTM 🚀  (not sure if we want to get a final 👍  from @jasmussen and/or @mirka )

**Comment by jasmussen** [comment]
I'll trust you all on the inner mechanisms here! What's important on the visual side is that the letter case/text transform remain visually as they appear [here](https://github.com/WordPress/gutenberg/pull/39760#issuecomment-1081861004). and aren't suddenly bordered or something.

**Pull request merged**
